examples/helia-browser-verified-fetch/src/utils.ts

```diff
@@ -0,0 +1,91 @@
+import { generateKeyPair } from '@libp2p/crypto/keys'
+import { createVerifiedFetch, VerifiedFetch } from '@helia/verified-fetch'
+import { createDelegatedRoutingV1HttpApiClient } from '@helia/delegated-routing-v1-http-api-client'
+import { createHelia } from 'helia'
+import * as libp2pInfo from 'libp2p/version'
+import { httpGatewayRouting, libp2pRouting } from '@helia/routers'
+import { enable } from '@libp2p/logger'
+import { createLibp2p, Libp2pOptions } from 'libp2p'
+import { webTransport } from '@libp2p/webtransport'
+import { noise } from '@chainsafe/libp2p-noise'
+import { yamux } from '@chainsafe/libp2p-yamux'
+import { dcutr } from '@libp2p/dcutr'
+import { identify, identifyPush } from '@libp2p/identify'
+import { keychain } from '@libp2p/keychain'
+import { ping } from '@libp2p/ping'
+import { webRTCDirect } from '@libp2p/webrtc'
+import { webSockets } from '@libp2p/websockets'
+
+// Singleton verifiedFetch instances
+let verifiedFetch: VerifiedFetch | undefined
+let verifiedFetchP2P: VerifiedFetch | undefined
+
+interface VerifiedFetchOptions {
+  useLibp2p?: boolean
+  useRecursiveGateways?: boolean
+}
+
+export async function getVerifiedFetch(
+  options: VerifiedFetchOptions = { useLibp2p: false, useRecursiveGateways: true },
+): Promise<VerifiedFetch> {
+  console.log('getVerifiedFetch', options)
+
+  enable('*')
+  if (!options.useLibp2p) {
+    verifiedFetch = verifiedFetch ?? (await createVerifiedFetch())
+    return verifiedFetch
+  }
+
+  verifiedFetchP2P = verifiedFetchP2P ?? (await createVerifiedFetchP2P(options))
+
+  console.log('verifiedFetchP2P', typeof verifiedFetchP2P)
+
+  return verifiedFetchP2P
+}
+
+async function createVerifiedFetchP2P(options: VerifiedFetchOptions): Promise<VerifiedFetch> {
+  const libp2pOptions = await libp2pDefaults()
+
+  const libp2p = await createLibp2p(libp2pOptions)
+
+  console.log('libp2p', libp2p)
+  console.log('options.useRecursiveGateways', options.useRecursiveGateways)
+  debugger
+  const helia = await createHelia({
+    libp2p,
+    routers: options.useRecursiveGateways
+      ? [httpGatewayRouting(), libp2pRouting(libp2p)]
+      : [libp2pRouting(libp2p)],
+  })
```

Comment:
I think it would be great if we could override routers with a callback that receives the libp2p instance so we didn't have to separately instantiate our own.. we could maybe change that in https://github.com/ipfs/helia/pull/652, but that would involve messing with the construction options quite a bit

Reply:
That's a good idea. Before I get into that I want to actually make sure we actually need libp2pRouting. 

Because we can also pass the DelegatedHTTPRouter. 

Which makes me wonder:

- If you have libp2p set up with only delegated routing, does it make sense to pass `libp2pRouting` as we do now, or better to directly use `delegatedHTTPRouting`? 
- What happens if you have both `delegatedHTTPRouting` and `libp2pRouting` with libp2p configured with delegated routing? Will it result in duplicate results?

Reply:
TL;DR: if we are using libp2p in Helia, best to use libp2p routing with the libp2p instance configured with the delegated router.